### ISSUE TITLE
[v4] A configuration object could not be found

### ISSUE THREAD

**Issue by cometkim** [angeloashmore/gatsby-source-prismic#384]
the error:
```
withPrismicPreview.tsx:109 Error: A configuration object could not be found for repository "karrot". Check that the repository is configured in your app's usePrismicPreviewResolver.
    at eval (usePrismicPreviewBootstrap.ts:150)
    at eval (Either.js:70)
    at eval (function.js:180)
    at pipe (function.js:277)
    at eval (usePrismicPreviewBootstrap.ts:145)
    at eval (Chain.js:5)
    at eval (EitherT.js:24)
    at eval (ReaderT.js:12)
    at eval (Task.js:110)
eval @ withPrismicPreview.tsx:109
```

`gatsby-config.js`

```
    {
      resolve: 'gatsby-source-prismic',
      options: {
        repositoryName: 'karrot',
        accessToken: process.env.PRISMIC_ACCESS_TOKEN,
        schemas: {
          'mbti-test-result': {},
          mbti_intro: {},
          mbti_test_question: {},
          ads_intro: {},
          global_contents: {},
          faq: require('@karrotmarket/prismic-config/schema/faq.json'),
          site_navigation: require('@karrotmarket/prismic-config/schema/site_navigation.json'),
          terms_and_conditions: require('@karrotmarket/prismic-config/schema/terms_and_conditions.json'),
          member_profile: require('@karrotmarket/prismic-config/schema/member_profile.json'),
          team_contents: require('@karrotmarket/prismic-config/schema/team_contents.json'),
        },
      },
    },
    {
      resolve: 'gatsby-plugin-prismic-previews',
      options: {
        repositoryName: 'karrot',
        accessToken: process.env.PRISMIC_ACCESS_TOKEN,
        toolbar: 'new',
      },
    },
```

wrapRootElement (both `gatsby-browser.jsx` and `gatsby-ssr.jsx`)

```
import * as React from 'react';
import type { GatsbyBrowser, GatsbySSR } from 'gatsby';
import { PrismicPreviewProvider } from 'gatsby-plugin-prismic-previews';

export const wrapRootElement: (GatsbyBrowser | GatsbySSR)['wrapRootElement'] = ({
  element,
}) => {
  return (
    <PrismicPreviewProvider>
      {element}
    </PrismicPreviewProvider>
  );
};
```

the preview resolver page

```
export const linkResolver: LinkResolver = doc => {
  switch (doc.type) {
    case 'site_navigation': {
      return '/';
    }
    case 'faq': {
      return '/faq/';
    }
    case 'team_contents': {
      return '/';
    }
  }
  return '/';
}

export default withPrismicPreviewResolver(PreviewResolverPage, [
  {
    repositoryName: 'karrot',
    linkResolver,
  },
]);
```

**Comment by cometkim** [comment]
the preview page is working as expected, redirect user to target page, provide the merged data into page.

**Comment by angeloashmore** [comment]
@cometkim Are you saying the preview works normally, but an error is printed to the log? Can you confirm which version of the plugin you are using (the latest is `4.0.0-beta.11` as of this comment)?

Thanks!

**Comment by greatwitenorth** [comment]
I'm seeing the same following error:

<img width="1015" alt="Screen Shot 2021-06-01 at 1 08 13 PM" src="https://user-images.githubusercontent.com/547583/120370636-77975e80-c2da-11eb-9858-03c1bc013439.png">

This appears to be happening on all previews whether they are published or drafts. The preview page only says loading, then displays the error.

This is what I'm seeing in the console:
<img width="1434" alt="Screen Shot 2021-06-01 at 1 16 44 PM" src="https://user-images.githubusercontent.com/547583/120371580-a3ffaa80-c2db-11eb-99ba-26d1cdd40071.png">

I'm using:
- Node v14.15.4
- Gatsby v3.6.1
-  gatsby-plugin-prismic-previews@4.0.0-beta.12
- gatsby-source-prismic@4.0.0-beta.11

**Comment by jodiedoubleday** [comment]
I'm also seeing the same error...

<img width="579" alt="Screenshot 2021-06-02 at 09 56 06" src="https://user-images.githubusercontent.com/778942/120450854-c7346380-c388-11eb-9172-0dabca96ff52.png">

**Comment by angeloashmore** [comment]
FYI, I'm working on this issue but I'm having a hard time reproducing it. If you have any ideas or notice anything strange, please share!

I'll post updates here as I learn more.

**Comment by cometkim** [comment]
@angeloashmore I will send invite you my working repo

**Comment by cometkim** [comment]
@angeloashmore I just sent invitation to https://github.com/daangn/websites

And I'm currently working with gatsby-source-prismic beta on https://github.com/daangn/websites/tree/main/team.daangn.com

I can generate a prismic preview link when you needed. at me or DM me on twitter anytime you want.

**Comment by angeloashmore** [comment]
Thank you @cometkim! I will take a look and reach out via Twitter if I get stuck.

**Comment by jodiedoubleday** [comment]
I can also add you to my repo if you need to compare, just give me a shout.

**Comment by angeloashmore** [comment]
@cometkim I was able to get previews working in your repo. Thanks for sharing it!

**TL;DR**: @greatwitenorth @jodiedoubleday @cometkim
Can you check that all pages using `withPrismicPreview`, `withPrismicUnpublishedPreview`, and `withPrismicPreviewResolver` have the repository configurations setup correctly?

## The problem

The preview resolver page was working fine. The Link Resolver was returning `/` for most documents rather than the correct URL. This is okay, it just means you'll land on the home page rather than the document's page. The homepage, however, did not have `withPrismicPreview` setup correctly.

Before (not working):
```js
export default withPrismicPreview(IndexPage, [])
```

After (working):
```js
export default withPrismicPreview(IndexPage, [
  {
    repositoryName: 'karrot',
    linkResolver
  },
])
```

I updated the error message to be a little more descriptive. Basically, all pages that should be previewable should have `withPrismicPreview` around the default export. This involves adding the repository config as well. If the higher order component doesn't find a configuration for the repository, it will display the following error.

<img src="https://user-images.githubusercontent.com/8601064/121106037-ebd17680-c7a0-11eb-8f23-082031265dbc.png" width="400" />

One way to simplify this is to make dedicated `prismicPreviews.ts` and `prismicUnpublishedPreviews.ts` files and setup your configuration objects there. Then you can import them into each page/template.

This repository contains an example of how I would recommend setting up previews: https://github.com/angeloashmore/gatsby-source-prismic/tree/beta/starters/gatsby-starter-prismic-typescript

**Examples**:
- `prismicPreviews.ts`: https://github.com/angeloashmore/gatsby-source-prismic/blob/beta/starters/gatsby-starter-prismic-typescript/src/prismicPreviews.ts
- `prismicUnpublishedPreviews.ts`: https://github.com/angeloashmore/gatsby-source-prismic/blob/beta/starters/gatsby-starter-prismic-typescript/src/prismicUnpublishedPreviews.ts
- Template: https://github.com/angeloashmore/gatsby-source-prismic/blob/beta/starters/gatsby-starter-prismic-typescript/src/pages/%7BPrismicPage.url%7D.tsx

**Comment by greatwitenorth** [comment]
@angeloashmore Thanks for looking into this. Unfortunately I've gone over my code multiple times and I cannot see any issues in the way I've implemented withPrismicUnpublishedPreview, withPrismicPreviewResolver and withPrismicPreview. They all contain the repository name and relevant configuration. I'm still seeing the same error (I've updated to beta15 as 16 doesn't seem to be available yet on npm).

Here's my template pages:

case_study.jsx:
```js
export default withPrismicPreview(CaseStudy, [
  {
    repositoryName: "weareroadmap",
    linkResolver
  },
])
```

insight.jsx:
```js
export default withPrismicPreview(Insight, [
  {
    repositoryName: "weareroadmap",
    linkResolver
  },
])
```

page:
```js
export default withPrismicPreview(Page, [
  {
    repositoryName: "weareroadmap",
    linkResolver
  },
])
```

preview:
```js
export default withPrismicPreviewResolver(PreviewPage, [
  {
    repositoryName: "weareroadmap",
    linkResolver,
    fetchLinks: [
      "page.slug_prefix",
      "case_study.slug_prefix",
      "insight.slug_prefix"
    ]
  }
])
```

404.js:
```js
export default withPrismicUnpublishedPreview(NotFoundPage, [
  {
    repositoryname: "weareroadmap",
    linkResolver,
    componentResolver: componentResolverFromMap({
      page: Page,
      insight: Insight,
      case_study: CaseStudy
    })
  }
])
```

I'd also be happy to share repo access, although it's hosted on bitbucket.

**Comment by jodiedoubleday** [comment]
I've got mine working by changing...
```
export default withPrismicPreview(PageTemplate, [
  {
    repositoryName: process.env.GATSBY_PRISMIC_MAIN_REPOSITORY_NAME,
    linkResolver,
  },
])
```

to

```
export default withPrismicPreview(PageTemplate, [
  {
    repositoryName: 'juniper-studio',
    linkResolver,
  },
])
```

**Comment by angeloashmore** [comment]
@greatwitenorth Thanks, I had a look at the code. Published previews worked fine for me, but I did need to add `_previewable` to the GraphQL queries. Without it, preview content won't be merged.

In `404.js`, however, there was small typo that was causing the red error modal to appear. Note the `repositoryName` casing:

```diff
  export default withPrismicUnpublishedPreview(NotFoundPage, [
    {
-     repositoryname: "weareroadmap-clone",
+     repositoryName: "weareroadmap-clone",
      linkResolver,
      componentResolver: componentResolverFromMap({
        page: Page,
        insight: Insight,
        case_study: CaseStudy,
      }),
    },
  ])
```

---

@jodiedoubleday Glad you got it working! In general, I would recommend sticking to the `process.env.GATSBY_PRISMIC_REPOSITORY_NAME` pattern as it keeps your configuration more portable. You'll need to make sure it is setup properly in a `.env` file as well as wherever you are building your site (Netlify for example).

That being said, providing your repository name as a string works as well!

**Comment by FagerKod** [comment]
@angeloashmore I'm also having trouble getting these previews to work, followed some instructions in the tutorial and got some help from you earlier with my package.json-versions. Mind taking a look att what I am missing?

https://github.com/FagerKod/prismiclab

**Comment by angeloashmore** [comment]
Hey @FagerKod, are you getting the same "A configuration object could not be found" error in the red modal?

Everything looks okay in your repository. The only unnecessary part I see is the Prismic toolbar in `Layout.js`. You can remove this as `gatsby-plugin-prismic-previews` will add it for you. This shouldn't be causing the problem, however.

https://github.com/FagerKod/prismiclab/blob/d9d0c835fda2999463c762b520d52952e63aa6c1/src/components/layout.js#L13-L17

**Comment by FagerKod** [comment]
@angeloashmore removed the script from Layout, thanks.
Still does not work, adding screen shot.

<img width="1005" alt="Skärmavbild 2021-06-14 kl  08 49 49" src="https://user-images.githubusercontent.com/19273692/121850918-c5a25e00-cced-11eb-93f3-a52a90f1f366.png">

**Comment by FagerKod** [comment]
Got my stuff working, had to import linkResolver in a correct way e.g
import linkResolver ...
instead of import {linkResolver} ...

Thanks for input @angeloashmore !

**Comment by angeloashmore** [comment]
Ah, glad you got it working! I'm learning that exact problem is actually quite common with the Link Resolver function. Maybe there's a better way we can teach how it set up the export/import.

**Comment by cometkim** [comment]
I tried again with the latest version of the plugin and got the same error.

https://github.com/daangn/websites/pull/121

![image](https://user-images.githubusercontent.com/9696352/123580176-3da37480-d814-11eb-846a-632aaccb730a.png)
![image](https://user-images.githubusercontent.com/9696352/123580235-5ad84300-d814-11eb-8073-20a9938b05a8.png)

**Comment by angeloashmore** [comment]
Hey @cometkim, it looks like your app doesn’t include PrismicPreviewProvider in `gatsby-browser.ts`.

Instructions: https://github.com/angeloashmore/gatsby-source-prismic/tree/beta/packages/gatsby-plugin-prismic-previews#gatsby-browser-and-ssr-files

Can you confirm this is in the app tree and let me know if the issue persists?

**Comment by cometkim** [comment]
@angeloashmore it is included in a theme I use

https://github.com/daangn/websites/pull/121/files#diff-9502760eb59a5f354737ca6093b53d94e5ea3083d1356b40099ee5b8acf5ea6a

**Comment by angeloashmore** [comment]
Got it, that makes sense. Can you try adding the provider directly in the site-specific gatsby-browser.ts file to see if it makes a difference? Trying to narrow down where the issue is happening. Thanks 👍

**Comment by cometkim** [comment]
Oh, sorry @angeloashmore  It was my bad. I just forgot to build my theme on my local 😅 

The previewing works fine in the latest version of the plugin. But it seems there's a SSR issue

![image](https://user-images.githubusercontent.com/9696352/123597474-591b7900-d82e-11eb-928f-63e694edb855.png)

The preview HoC doesn't seems to guarantee isomorphic results, the hydration is not work properly and layouts that rely on media queries look broken.

**Comment by cometkim** [comment]
Maybe I can close this and open another issue

**Comment by angeloashmore** [comment]
Hey @cometkim, no problem! Glad it was an easy fix. Yes, I'll close this issue but if anyone else runs into configuration issues, feel free to post here and I can help.

**Comment by danspratling** [comment]
@angeloashmore I've found a small documentation discrepency that led to me getting this error for a while.

Partially my fault for copy/pasting without reading fully, but something many others can/will run into too.

Here, the API key is `GATSBY_PRISMIC_*REPO*_NAME`
https://prismic.io/docs/technologies/install-the-plugin-gatsby

And here it's both `GATSBY_PRISMIC_*REPO*_NAME` and `GATSBY_PRISMIC_*REPOSITORY*_NAME` depending on which part of the page you're looking at.
https://prismic.io/docs/technologies/previews-gatsby

**Comment by angeloashmore** [comment]
@danspratling Good catch! That error will be fixed shortly.

**Comment by ErikAGriffin** [comment]
Adding this context for someone else running into a similar issue in the future:

For me I narrowed down the issue as coming from the `gatsby-browser.js` file, and that replacing the `process.env.<repo-name>` variable with a string fixed it.  My PrismicProvider looked something like this:
```javascript
export const wrapRootElement = ({ element }) => (
  <PrismicProvider
    internalLinkComponent={({ href, ...props }) => (
      <Link to={href} {...props} />
    )}
  >
    <PrismicPreviewProvider
      repositoryConfigs={[
        {
          repositoryName: process.env.PRISMIC_REPO_NAME,
          linkResolver,
          componentResolver: componentResolverFromMap({
            front_page: FrontPage,
          }),
        },
      ]}
    >
      {element}
    </PrismicPreviewProvider>
  </PrismicProvider>
)
```
Can you spot the issue?  Depending on what framework you are using, environment variables are not allowed to be accessed by the client browser, *unless* they are prefixed with a particular string.  So for me, changing the environment variable to `GATSBY_PRISMIC_REPO_NAME` solved the issue.
